Route based annotations for service accounts

enj opened the pull request:
Allows the specification of a route as a valid redirect URI.  Thus all ingress hosts for that route would be considered valid.

cc @gabemontero @bparees 

Trello ref:
https://trello.com/c/fKEDAntM

Minimal annotations needed to dynamically pick up a redirect URI from a route:

```json
{
  "kind": "ServiceAccount",
  "apiVersion": "v1",
  "metadata": {
    "name": "jenkins",
    "annotations": {
      "serviceaccounts.openshift.io/oauth-redirectreference.1": "{\"kind\":\"OAuthRedirectReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"Route\",\"name\":\"jenkins\"}}"
    }
  }
}
```

enj commented:
[test]

mfojtik commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -10,27 +12,88 @@ import (
+}
+
+func (uri *redirectURI) merge(m model) {
+	if m.scheme != "" {
```
len(m.scheme) == 0

enj replied:
Any reason to prefer `len` vs. comparison to `""`?  I find the latter more specific as it denotes that the thing being compared must be a `string`, whereas `len` works for anything that can hold items.

liggitt replied:
convention

mfojtik replied:
@enj yeah, we picked len() convention so we can nit pick people pr's ;-)

mfojtik commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -10,27 +12,88 @@ import (
+		host = uri.host
+	}
+	u, err := url.Parse(fmt.Sprintf("%s://%s/%s", uri.scheme, host, uri.path))
+	if err == nil && u.Fragment == "" {
```
len()

mfojtik commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -10,27 +12,88 @@ import (
+
+func (uri *redirectURI) getURI() (string, error) {
+	var host string
+	if uri.port != "" {
```
len()

mfojtik commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -10,27 +12,88 @@ import (
+func (uri *redirectURI) getURI() (string, error) {
+	var host string
+	if uri.port != "" {
+		host = fmt.Sprintf("%s:%s", uri.host, uri.port)
```
host = uri.host + ":" + uri.port

mfojtik commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -81,6 +154,121 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
 	return saClient, nil
 }
 
+func updateModels(models map[string]model, key, value string) {
```
i think this deserves some refactoring :-)

enj replied:
Any specific suggestions?

liggitt replied:
```
func parseModelPrefixName(key string) (string, string, bool) {
  for _, prefix := range modelPrefixes {
    if strings.HasPrefix(key, prefix) {
      return prefix, key[len(prefix):], true
    }
  }
  return false
}

func updateModels(models map[string]model, key, value string) {
  prefix, name, ok := parseModelPrefixName(key)
  if !ok {
    return
  }
  m = models[name]
  switch prefix {
    case OAuthRedirectModelAnnotationURISchemePrefix:
      m.scheme = value
    ...
  }
  models[name] = m
}
```

mfojtik commented:
@liggitt can you have a very quick look?

liggitt commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -10,27 +12,88 @@ import (
+}
+
+func (uri *redirectURI) getURI() (string, error) {
+	var host string
```
```
host := uri.host
if len(uri.port) > 0 {
  host = net.JoinHostPort(host, uri.port)
}
return &url.URL{Scheme:uri.scheme, Host: host, Path: uri.path}.String()

enj replied:
Should we validate that the scheme is either `http` or `https`?

enj replied:
The general question being is that do we need to validate that things like the `port`, `scheme`, etc make sense.

liggitt commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -10,27 +12,88 @@ import (
+	}
+	if m.port != "" {
+		uri.port = m.port
+	}
```
host?

enj replied:
I populate the defaults / known values in `updateRouteMap` - haven't figured out how to tell from the route data if the route wants to use TLS or not.

liggitt replied:
merge should merge host as well

enj replied:
Not sure what you mean here since `model` does not have a host field.

liggitt replied:
it should, don't special-case the host portion of the uri. build the stack of models (defaults, annotation bits, object bits), merge them, and take the resulting uri

enj commented:
Still trying to figure out what to call the base constant, thoughts are:

1. serviceaccounts.openshift.io/oauth-redirect**template**.
2. serviceaccounts.openshift.io/oauth-redirect**model**.
3. serviceaccounts.openshift.io/oauth-redirect**rule**.

liggitt commented on pkg/cmd/server/origin/master_config.go:
```diff
@@ -790,6 +790,11 @@ func (c *MasterConfig) KubeClient() *kclient.Client {
 }
 
+// OpenShiftClient returns the openshift client object
+func (c *MasterConfig) OpenShiftClient() *osclient.Client {
```
I'd rather have an `OAuthServerClients() (*osclient.Client, *kclient.Client)` function to clarify what the returned clients are to be used for

mfojtik replied:
also maybe highlight that this will return the privileged client? btw. why we need this function if it is only returning a field that is already public?

enj replied:
I believe the functions just exist to document how the client will be used.  The `*Client` functions essentially always return those public fields.

liggitt replied:
actually, a lot of them return clients backed by service account credentials

enj replied:
OK, should I document that the clients returned are privileged?

mfojtik commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -81,6 +161,117 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
+}
+
+func isOpenShiftRoute(m *model) bool {
+	return m.kind == RouteKind && m.group == routeapi.FutureGroupName
```
@deads2k is using `FutureGroupName` kosher?

deads2k replied:
> @deads2k is using FutureGroupName kosher?

It probably won't do what you want.  Is someone using it elsewhere?  I probably wouldn't do that yet.

liggitt commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -10,27 +12,112 @@ import (
+	OAuthRedirectModelAnnotationResourceGroupPrefix,
+}
+
+type modelsToURIsFunc func(modelList []model) []redirectURI
```
could this be something that would take a set of names for a particular groupkind, and return the components for the objects with those names? something like:
```
func uriComponentsFromRoutes(names sets.String) map[string][]uriComponents {
   var routes []api.Route 
   if len(names) == 1 {
     ... look up one route ...
   }else {
     ... list routes ...
   }

   retval := map[string][]uriComponents{}
   for _, route := range routes {
     if names.Has(route.Name) {
       retval[route.Name] = uriComponentsFromRoute(route)
     }
   }
   return retval
}

func uriComponentsFromRoute(route api.Route) []uriComponents {
  scheme := "http"
  if route.TLS != nil {
    scheme = "https"
  }

  retval := []uriComponents{}
  for _, ingress := route.Status.Ingress {
    if len(ingress.Host) > 0 && ... {
      retval = append(retval, uriComponents{Scheme: scheme, Host: ingress.Host}
    }
  }
  return retval
}

```

enj commented:
re[test] flake zombie #10002, possibly needs triage #7165

enj commented:
re[test] flake again zombie #10002, possibly needs triage #7165

soltysh commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -73,14 +203,117 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
+
+	for gk, models := range groupKindModelListMapper {
+		if len(models) > 0 {
+			if names := models.getNames(); names.Len() > 0 {
```
These two ifs are checking the same think, merge them together. Eventually `getNames` function should have a short path for when models len is 0.

enj replied:
They don't check the same thing; it is possible for `len(models) > 0` and `names.Len() == 0`.

soltysh replied:
ok

soltysh commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -73,14 +203,117 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
+	routeInterface := a.routeClient.Routes(namespace)
+	if osRouteNames.Len() > 1 {
+		r, err := routeInterface.List(kapi.ListOptions{})
+		if err == nil {
```
I'm a big fan of short-if, ask @pweil- :wink:, he calls me short-if :cop::
```
if r, err := routeInterface.List(kapi.ListOptions{}); err == nil {
...
```

soltysh commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -73,14 +203,117 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
+		}
+	} else {
+		r, err := routeInterface.Get(osRouteNames.List()[0])
+		if err == nil {
```
Same.

enj commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -10,27 +12,150 @@ import (
+	return data
+}
+
+func (uri *redirectURI) merge(m model) {
```
Eh, `merge` makes more sense to me since I am merging all the useful information in `model` with the `redirectURI`.

soltysh replied:
ok

enj commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -73,14 +203,117 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
+func parseModelsMap(annotations map[string]string) map[string]model {
+	models := map[string]model{}
+	for key, value := range annotations {
+		if prefix, name, ok := parseModelPrefixName(key); ok {
```
Wouldn't this be inconsistent with being the short-if :cop: ?  I find the use of a `continue` to be less clear...

soltysh replied:
Only slightly a bit, readability is what matters most :smile:

soltysh commented:
LGTM on my side, squash and we need final sign-off from @liggitt before merging this.

enj commented:
cc @gabemontero @bparees in case they have any comments.

gabemontero commented:
No immediate comments.  My hope/intent is to checkout this branch locally
and try it out in the short term unless it gets through the merge queue
beforehand.

On Thu, Sep 22, 2016 at 3:43 PM, Mo Khan <notifications@github.com> wrote:

> cc @gabemontero <https://github.com/gabemontero> @bparees
> <https://github.com/bparees> in case they have any comments.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/10878#issuecomment-249007498>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbadLQ_1tL5mrwuX6O8eWjrOQyCnWjoks5qstpNgaJpZM4J6nbN>
> .
>

liggitt commented:
sample usage in the description?

soltysh commented:
> sample usage in the description?

I'm hoping for a good docs for this :smile:

enj commented:
I added an example service account above; it should be enough for someone to test with.

mfojtik commented:
@openshift/api-review  need a check of the name

smarterclayton commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -10,27 +12,150 @@ import (
 	OAuthRedirectURISecretAnnotationPrefix = "serviceaccounts.openshift.io/oauth-redirecturi."
 	OAuthWantChallengesAnnotationPrefix    = "serviceaccounts.openshift.io/oauth-want-challenges"
+
+	OAuthRedirectModelAnnotationPrefix              = "serviceaccounts.openshift.io/oauth-redirectmodel."
```
What is a "redirect model"?  We don't use the word "model" in our public apis today, so would need a good justification.  Explain meaning so we can find a good match.

enj replied:
So I dislike all of the names I have come up with for this API, so I welcome suggestions.  That being said, a redirect model refers to the following fields that could be used to build redirect URI(s):
```
uri.scheme
uri.port
uri.path
uri.host
resource.kind
resource.name
resource.group
```
The `resource` components define where to get the default redirect data from (route being the only resource we support for now).  If specified, the `uri` components are used to override the default data.  As long as the resulting URI(s) have a scheme and a host, they are considered valid.

smarterclayton commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -10,27 +12,150 @@ import (
+	OAuthRedirectModelAnnotationResourceGroupPrefix,
+}
+
+type namesToObjMapperFunc func(namespace string, names sets.String) map[string]redirectURIList
```
Serious lack of Godoc in this file explaining intent / algo

smarterclayton commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -10,27 +12,150 @@ import (
+	OAuthRedirectModelAnnotationPrefix              = "serviceaccounts.openshift.io/oauth-redirectmodel."
+	OAuthRedirectModelAnnotationURIPrefix           = OAuthRedirectModelAnnotationPrefix + "uri."
+	OAuthRedirectModelAnnotationResourcePrefix      = OAuthRedirectModelAnnotationPrefix + "resource."
+	OAuthRedirectModelAnnotationURISchemePrefix     = OAuthRedirectModelAnnotationURIPrefix + "scheme."
```
Are these partial overrides?  Why aren't they on the route instead?

enj replied:
These are partial overrides on redirect URI data fetched from some resource (which may or may not be a route).  The annotations are on a service account because that account is stating what redirect URIs are valid for it - the other direction does not make as much sense.  Plus it goes against the existing `serviceaccounts.openshift.io/oauth-redirecturi.` annotation.

kargakis commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -73,14 +203,117 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
+			routes = r.Items
+		}
+	} else {
+		if r, err := routeInterface.Get(osRouteNames.List()[0]); err == nil {
```
Are you sure the provided names are always positive (>0)?

enj replied:
`getNames` is used to build the `sets.String` and it filters out empty strings.  Thus at this point in the code we are guaranteed to have at least one item in the set and all items will be non-empty strings.

kargakis commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -73,14 +203,117 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
 	return saClient, nil
 }
 
+func parseModelsMap(annotations map[string]string) map[string]model {
```
most of these helpers need godoc

kargakis commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -73,14 +203,117 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
 }
 
+func parseModelsMap(annotations map[string]string) map[string]model {
+	models := map[string]model{}
```
nit: use `var models map[string]model` to avoid the extra allocation

enj replied:
IMHO the overhead of dealing with a `nil`

enj commented:
@smarterclayton 
> When we add ingress support, how will that work?

We will add a new `GroupKind` combination for `Ingress` and create a new  `namesToObjMapperFunc` similar to `redirectURIsFromRoutes`.  The external annotations API will remain unchanged, just more annotations would become valid.

smarterclayton commented:
Add the ingress statement as a future looking statement someplace relevant in your code.

liggitt commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -10,27 +12,175 @@ import (
+// These values can be overridden by user specified data.
+type namesToObjMapperFunc func(namespace string, names sets.String) map[string]redirectURIList
+
+var routeGroupKind = unversioned.GroupKind{Group: routeapi.FutureGroupName, Kind: RouteKind}
```
I would start by supporting the current (`""`) group, and later add support for the openshift.io group:
```
var routeGroupKind = routeapi.SchemeGroupVersion.WithKind("Route").GroupKind()
```

enj commented:
Addressed new comments and squashed.  Waiting on @liggitt to review, @smarterclayton to approve, and @gabemontero to use.  Name suggestions for the annotation are welcome from all.

gabemontero commented:
I have been able to consume the changes from this PR's branch and verify that I can perform the oauth login/logout flow from Jenkins with the route defined by the openshift example jenkins template without having to manually annotate the jenkins service account.

smarterclayton commented:
I'm not in favor of model, try to find something more consistent with
existing API usage

On Thu, Sep 29, 2016 at 3:10 PM, Gabe Montero <notifications@github.com>
wrote:

> I have been able to consume the changes from this PR's branch and verify
> that I can perform the oauth login/logout flow from Jenkins with the route
> defined by the openshift example jenkins template without having to
> manually annotate the jenkins service account.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/10878#issuecomment-250562419>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p7IcSq1k90Io5NueKCpODpMa13Lsks5qvA0WgaJpZM4J6nbN>
> .
>

liggitt commented:
existing annotation:
```
"serviceaccounts.openshift.io/oauth-redirecturi.foo":"https://myhost"
```

what about
```
"serviceaccounts.openshift.io/oauth-redirectobjectkind.foo": "Route"
"serviceaccounts.openshift.io/oauth-redirectobjectname.foo": "jenkins"
```

if you needed an API group:
```
"serviceaccounts.openshift.io/oauth-redirectobjectkind.foo": "Ingress"
"serviceaccounts.openshift.io/oauth-redirectobjectgroup.foo": "extensions"
"serviceaccounts.openshift.io/oauth-redirectobjectname.foo": "jenkins"
```

if you needed to customize path, port, etc:
```
"serviceaccounts.openshift.io/oauth-redirecturi.foo":"https://:8443/custom/path"
"serviceaccounts.openshift.io/oauth-redirectobjectkind.foo": "Route"
"serviceaccounts.openshift.io/oauth-redirectobjectname.foo": "jenkins"
```

enj commented:
@gabemontero I have updated the code to follow Jordan's recommendations - you will have to update your template to match the new keys (granted they aren't set in stone until @smarterclayton agrees).

smarterclayton commented:
Is there a reason to not have more internal dashes to separate the words?

enj commented:
I suppose there are a few ways to do it:

```
"serviceaccounts.openshift.io/oauth-redirecturi.foo":"https://:8443/custom/path"
"serviceaccounts.openshift.io/oauth-redirectobjectkind.foo": "Route"
"serviceaccounts.openshift.io/oauth-redirectobjectname.foo": "jenkins"
"serviceaccounts.openshift.io/oauth-redirectobjectgroup.foo": ""
```

```
"serviceaccounts.openshift.io/oauth-redirecturi.foo":"https://:8443/custom/path"
"serviceaccounts.openshift.io/oauth-redirect-object-kind.foo": "Route"
"serviceaccounts.openshift.io/oauth-redirect-object-name.foo": "jenkins"
"serviceaccounts.openshift.io/oauth-redirect-object-group.foo": ""
```

```
"serviceaccounts.openshift.io/oauth-redirecturi.foo":"https://:8443/custom/path"
"serviceaccounts.openshift.io/oauth-redirectobject-kind.foo": "Route"
"serviceaccounts.openshift.io/oauth-redirectobject-name.foo": "jenkins"
"serviceaccounts.openshift.io/oauth-redirectobject-group.foo": ""
```

Just depends on how many dashes you want...

smarterclayton commented:
Given that we typically encode structured data as JSON, I think this needs to be:

```
"serviceaccounts.openshift.io/oauth-redirecturi.foo":"https://:8443/custom/path"
"serviceaccounts.openshift.io/oauth-redirectkind.foo": "{\"kind\": \"Route\", \"name\": \"jenkins\"}"
```

The serialization object should be `v1.ObjectReference`

enj commented:
@smarterclayton How would `v1.ObjectReference` work in regards to having a `group`?

liggitt commented:
> @smarterclayton How would `v1.ObjectReference` work in regards to having a `group`?

poorly. you would have to parse it out of the `APIVersion` field (which holds a value like `extensions/v1beta1`), and discard the version portion of it.

we'd also disallow the namespace field... can't pull from a route in another namespace

enj commented:
@liggitt @smarterclayton Can we decide on what we actually want to do?  I don't want to implement the `v1.ObjectReference` stuff and change all the tests to match just to have to undo it because we decide we want something else...

smarterclayton commented:
Thought it's pretty clear that we have an API pattern for structured data
in annotations.  Whether you use object reference (everyone else does) or
create an object reference lite is less important.

On Oct 5, 2016, at 2:43 PM, Mo Khan <notifications@github.com> wrote:

@liggitt <https://github.com/liggitt> @smarterclayton
<https://github.com/smarterclayton> Can we decide on what we actually want
to do? I don't want to implement the v1.ObjectReference stuff and change
all the tests to match just to have to undo it because we decide we want
something else...

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/10878#issuecomment-251762604>,
or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABG_pwsKxqES1LSf7HgUrPyXiJfGrqfLks5qw--9gaJpZM4J6nbN>
.

enj commented:
New format:

```
"serviceaccounts.openshift.io/oauth-redirecturi.foo":"https://:8443/custom/path"
"serviceaccounts.openshift.io/oauth-redirectref.foo": "{\"kind\": \"Route\", \"name\": \"jenkins\", \"group\": \"\"}"
```

smarterclayton commented:
API approved unless anyone has additional arguments to deploy.

marun commented:
flake #11374 re-[test]

enj commented:
@liggitt any final comments on the code and/or the API?

liggitt commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -10,27 +13,188 @@ import (
 }
 
+// redirectObjectReference holds the resource components needed to reference an object in the same namespace
+type redirectObjectReference struct {
```
if we're encoding this to annotations, we should follow the pattern of SerializedReference/ObjectReference:

1. you'll need two API types - one for the data you care about (group/kind/name), and one that is registered as a top-level type (with TypeMeta, so we can encode/decode the data in the annotation using standard object encoders/decoders)
2. generate the annotation by encoding the top-level object (see how createdByRefJson is created upstream)
3. parse the annotation by decoding the annotation into an object (see how creatorRefRaw is decoded into an object upstream)

enj commented:
Even newer format:

```json
"serviceaccounts.openshift.io/oauth-redirecturi.foo": "https://:8443/custom/path"
"serviceaccounts.openshift.io/oauth-redirectref.foo": "{\"kind\":\"SerializedRedirectReference\",\"apiVersion\":\"v1\",\"reference\":{\"group\":\"\",\"kind\":\"Route\",\"name\":\"jenkins\"}}"
```

@liggitt Let me know if this is good so I can squash.

liggitt commented:
API looks as good as serialized JSON in an annotation is going to look. @deads2k / @smarterclayton, do you want to keep the object type specific (`SerializedRedirectReference` or other better name) or would we use this for other things and pick a generic name (`ObjectReferenceDoneRight`)?

smarterclayton commented:
Specific name for now.  Either way we support this for a long time.

On Oct 20, 2016, at 9:03 AM, Jordan Liggitt <notifications@github.com>
wrote:

API looks as good as serialized JSON in an annotation is going to look.
@deads2k <https://github.com/deads2k> / @smarterclayton
<https://github.com/smarterclayton>, do you want to keep the object type
specific (SerializedRedirectReference or other better name) or would we use
this for other things and pick a generic name (ObjectReferenceDoneRight)?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/10878#issuecomment-255099610>,
or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p5ovmWmATgroj9rhofpUvERU72iRks5q12asgaJpZM4J6nbN>
.

smarterclayton commented:
Probably prefer the more specific kind now

On Oct 20, 2016, at 9:03 AM, Jordan Liggitt <notifications@github.com>
wrote:

API looks as good as serialized JSON in an annotation is going to look.
@deads2k <https://github.com/deads2k> / @smarterclayton
<https://github.com/smarterclayton>, do you want to keep the object type
specific (SerializedRedirectReference or other better name) or would we use
this for other things and pick a generic name (ObjectReferenceDoneRight)?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/10878#issuecomment-255099610>,
or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p5ovmWmATgroj9rhofpUvERU72iRks5q12asgaJpZM4J6nbN>
.

liggitt commented on pkg/api/validation/register.go:
```diff
@@ -1,8 +1,6 @@
 package validation
 
 import (
-	_ "github.com/openshift/origin/pkg/api/install"
```
why remove this?

enj replied:
It's imported twice which my IDE complains about / gets a bit confused by.  See the last import in that block.  Or perhaps there was some nuanced Go reason for doing this?

deads2k commented:
> Specific name for now.  Either way we support this for a long time.

Agree.  Narrow name for a narrow case.

liggitt commented on pkg/oauth/api/types.go:
```diff
@@ -167,3 +167,14 @@ type OAuthClientAuthorizationList struct {
 	Items []OAuthClientAuthorization
 }
+
+type SerializedRedirectReference struct {
```
to avoid name clashes, make this `OAuthRedirectReference`, and make the annotation match `oauth-redirectreference`

liggitt commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -2,35 +2,194 @@ package oauthclient
+	// Prefix used for dynamically specifying redirect URIs using resources for a service account via annotations
+	OAuthRedirectModelAnnotationReferencePrefix = OAuthRedirectModelAnnotationPrefix + "ref."
+
+	RouteKind = "Route"
```
don't export

liggitt commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -2,35 +2,194 @@ package oauthclient
+// getRedirectURIs uses the mapping provided by a namesToObjMapperFunc to enumerate all of the redirect URIs
+// based on the name of each resource.  The user provided data in the model overrides the data in the mapping.
+// The returned redirect URIs may contain duplicate and invalid entries.
+func (ml modelList) getRedirectURIs(objMapper map[string]redirectURIList) redirectURIList {
```
note that all the items in the modelList must have a uniform group/kind, and the objMapper must be specifically for that group/kind

liggitt commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -2,35 +2,194 @@ package oauthclient
+		if uris, ok := objMapper[m.name]; ok {
+			for _, uri := range uris {
+				u := uri
+				u.merge(&m)
```
will this mutate the item held by objMapper? testcase like this would be good:

```
serviceaccounts.openshift.io/oauth-redirecturi.1=https://:myport
serviceaccounts.openshift.io/oauth-redirectref.1=<ref to route foo>
serviceaccounts.openshift.io/oauth-redirecturi.2=https:///mypath
serviceaccounts.openshift.io/oauth-redirectref.2=<ref to route foo>
```

make sure we get back two distinct redirect uris, one with a custom port, one with a custom path

liggitt commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -73,14 +235,125 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
 }
 
+// parseModelsMap builds a map of model name to model using a service account's annotations.
+// The model name is only used for building the map and serves no functional purpose other than making testing easier.
```
the model name ties together redirecturi and redirectref annotations

liggitt commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -73,14 +235,125 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
+	uri := newDefaultRedirectURI()
+	uri.path = route.Spec.Path
+	if route.Spec.Port != nil {
+		uri.port = route.Spec.Port.TargetPort.String()
```
this isn't right... the target port determines what port on the endpoint the router routes to, not what port the route listens on. we don't actually have that information represented in the route today, so this should always make the port the default for the scheme ("")

liggitt commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -73,14 +235,125 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
+		uri.port = route.Spec.Port.TargetPort.String()
+	}
+	if route.Spec.TLS == nil {
+		uri.scheme = "http"
```
splitting this logic between newDefaultRedirectURI() and here is hard to follow... an else that sets `"https"` would be better

liggitt commented on pkg/serviceaccounts/oauthclient/oauthclientregistry.go:
```diff
@@ -73,14 +235,125 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
+	}
+	for _, ingress := range route.Status.Ingress {
+		u := uri
+		u.host = ingress.Host
```
only if `len(ingress.Host) > 0`

liggitt replied:
follow up to only include this if the condition is that the route was accepted by the router?

enj replied:
The test `host overrides route data` would fail if I did that check here instead of letting `extractValidRedirectURIStrings` catch it later.

liggitt replied:
hmm, worth a comment then

liggitt commented on pkg/oauth/api/validation/validation.go:
```diff
@@ -294,3 +294,19 @@ func ValidateScopes(scopes []string, fldPath *field.Path) field.ErrorList {
 	return allErrs
 }
+
+func ValidateSerializedRedirectReference(sref *api.SerializedRedirectReference) field.ErrorList {
```
pass the object through ValidateObjectMeta

enj replied:
Are we expecting users to actually fill out any of the fields?  Won't this require them to give us a `metadata.Name`?

liggitt replied:
no, just to satisfy the validation test. if we ever persisted these, we'd run them through validation

enj replied:
Wouldn't it be easier to just exclude the object from the validation test?  Also, @deads2k wanted to use the validation to give a useful error message when the user tries to annotate a SA - problem is that SA stuff is upstream and wouldn't know about a Route.

marun commented:
networking job flaked #11452

liggitt commented:
LGTM, squash before merge

enj commented:
@mfojtik Ready for merge.

mfojtik commented:
[merge]

gabemontero commented:
I've pulled down the changes since this PR merged, but am having trouble getting the oauth whitelist properly updated from the serviceaccount annotations.

I'm getting messages like this in the master log:

```
E1021 23:32:08.938142   11055 osinserver.go:99] internal error: system:serviceaccount:myproject:jenkins has no redirectURIs; set serviceaccounts.openshift.io/oauth-redirecturi.<some-value>=<redirect> or create a dynamic URI using serviceaccounts.openshift.io/oauth-redirectreference.<some-value>=<reference>
```

Based on the last update I saw in the PR, my serviceaccount annotaion produced the following yaml output:
```
serviceaccounts.openshift.io/oauth-redirectref.jenkins:'{"kind":"SerializedRedirectReference","apiVersion":"v1","reference":{"group":"","kind":"Route","name":"jenkins"}}'
```

Then, based on the error log I noted above, I switched to:
```
serviceaccounts.openshift.io/oauth-redirectreference.jenkins:'{"kind":"SerializedRedirectReference","apiVersion":"v1","reference":{"group":"","kind":"Route","name":"jenkins"}}'
```

Here is the `oc get routes` output:

```
NAME      HOST/PORT                              PATH      SERVICES   PORT      TERMINATION
jenkins   jenkins-myproject.192.168.0.8.xip.io             jenkins    <all>     edge/Redirect
```

@enj - any pointers?

liggitt commented:
The example in the pr description is correct:

OAuthRedirectReference

gabemontero commented:
That did the trick - thanks @liggitt !!

I was able to log into Jenkins via the oauth plugin without any manual annotations to the SA using the Route for Jenkins and specifying the OAuthRedirectReference annotation in the jenkins template.
